### ISSUE TITLE
Is there an example for the YOLOv8 transfer Learning with custom dataset?

### ISSUE THREAD

**Issue by jshin10129** [ultralytics/ultralytics#8876]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

I've done almost every ways to make the transfer learning, but failed. Is there any example command and dataset (having only one new label at least) ?
According to "https://github.com/ultralytics/ultralytics/issues/6087", I'm supposed to keep the pretrained weights ("person") but can add a new label ("ball"). However, the solution here did not work.

Here is what I did:
***
yolo detect train data=test.yaml model=yolov8n.yaml pretrained=yolov8n.pt epochs=25 imgsz=640 plots=True

The logs are like this:
Transferred 319/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
...
YOLOv8n summary (fused): 168 layers, 3160775 parameters, 0 gradients, 8.8 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00
                   all          5         18          0          0          0          0
***
The thing is that I just added one label, "barrel" in the test.yaml.
And, just added one image under train and one image under val with labels text files.
Here is my test.yaml (as you see, 80:barrel) :
***
path: ../datasets/test
train: images/train
val: images/val 

names:
  0: person
  1: bicycle
  2: car
  3: motorcycle
  4: airplane
  5: bus
  6: train
  7: truck
  8: boat
  9: traffic light
  10: fire hydrant
  11: stop sign
  12: parking meter
  13: bench
  14: bird
  15: cat
  16: dog
  17: horse
  18: sheep
  19: cow
  20: elephant
  21: bear
  22: zebra
  23: giraffe
  24: backpack
  25: umbrella
  26: handbag
  27: tie
  28: suitcase
  29: frisbee
  30: skis
  31: snowboard
  32: sports ball
  33: kite
  34: baseball bat
  35: baseball glove
  36: skateboard
  37: surfboard
  38: tennis racket
  39: bottle
  40: wine glass
  41: cup
  42: fork
  43: knife
  44: spoon
  45: bowl
  46: banana
  47: apple
  48: sandwich
  49: orange
  50: broccoli
  51: carrot
  52: hot dog
  53: pizza
  54: donut
  55: cake
  56: chair
  57: couch
  58: potted plant
  59: bed
  60: dining table
  61: toilet
  62: tv
  63: laptop
  64: mouse
  65: remote
  66: keyboard
  67: cell phone
  68: microwave
  69: oven
  70: toaster
  71: sink
  72: refrigerator
  73: book
  74: clock
  75: vase
  76: scissors
  77: teddy bear
  78: hair drier
  79: toothbrush
  80: barrel
***
Is my test.yaml incorrect? or something wrong?

BTW, I'm using 640x640 images for the new label.
Is this image dimension matter?

Thanks for any helps in advance.

**Comment by glenn-jocher** [comment]
@jshin10129 hi there! 👋

It looks like you're on the right track with transfer learning using YOLOv8. Your `test.yaml` seems correctly set up for adding a new label "barrel". The image dimension of 640x640 is also fine as long as it matches the `imgsz` parameter you're using during training.

One thing to check is the format of your label files in the `train` and `val` folders. Ensure they match the YOLO format, where each line in a label file corresponds to a single object and follows the format `<class> <x_center> <y_center> <width> <height>`, all normalized to values between 0 and 1 relative to the image size.

Also, since you're adding a new class, make sure your dataset (even if small) has representative examples of this new class and that they are correctly labeled. This will help the model learn the new class effectively.

If you're still facing issues, consider using a subset of the COCO dataset that includes your new class "barrel" alongside the existing classes, to ensure there's enough data for effective learning.

Lastly, remember that training on a very small dataset might not yield significant results, so if possible, try to include more images for the new class.

Keep experimenting, and don't hesitate to reach out if you have more questions! 🚀

**Comment by glenn-jocher** [comment]
@chaitanya-srikanth hey there! 👋

It sounds like you're facing a challenge with transfer learning and introducing a new class to your pretrained model. When you're freezing layers and adding a new class, it's crucial to ensure that the later, unfrozen layers of the model have enough capacity to learn the new class' features. 

Here's a quick tip: Instead of freezing a fixed number of initial layers, you might want to experiment with freezing layers based on the model architecture. For example, in YOLO models, you could freeze up to the last convolutional layer before the YOLO layers start. This way, the model retains its ability to detect previously learned classes while adapting to the new class with the unfrozen layers.

```bash
# Example of specifying which layers to freeze in a hypothetical training command
yolo detect train data=my_dataset.yaml model=my_pretrained_model.pt freeze=[0-15]
```

Also, ensure your new dataset with the added class is well-represented and balanced. Sometimes, adding more varied examples of the new class can significantly help the model learn better.

Keep experimenting with the number of layers you freeze and the composition of your dataset. Sometimes, finding the right balance requires a bit of trial and error. 🚀

Let us know how it goes!

**Comment by jshin10129** [comment]
Hi,

I've tried all the suggested approaches from this issue thread, but still I didn't get the correct prediction outputs.
At this moment, I'm a bit puzzled about the transfer learning capability -- maybe there is a way to add another fully-connected layer (for learning new labels) on top of the pretrained layer? or do I need to tweak the argument values like "freeze=?" or "epoches=?" or "imgsz=?"

Here is my command to train:
`yolo detect train epochs=100 data=my_dataset.yaml model=yolov8s.pt imgsz=640 freeze=[1-30]`

And, here is "my_dataset.yaml":
`train: ./train
val: ./val

nc: 82

names: [
    "person", "bicycle", "car", "motorbike", "aeroplane", "bus", "train", "truck",
    "boat", "traffic light", "fire hydrant", "stop sign", "parking meter", "bench",
    "bird", "cat", "dog", "horse", "sheep", "cow", "elephant", "bear", "zebra",
    "giraffe", "backpack", "umbrella", "handbag", "tie", "suitcase", "frisbee",
    "skis", "snowboard", "sports ball", "kite", "baseball bat", "baseball glove",
    "skateboard", "surfboard", "tennis racket", "bottle", "wine glass", "cup",
    "fork", "knife", "spoon", "bowl", "banana", "apple", "sandwich", "orange",
    "broccoli", "carrot", "hot dog", "pizza", "donut", "cake", "chair", "sofa",
    "pottedplant", "bed", "diningtable", "toilet", "tvmonitor", "laptop", "mouse",
    "remote", "keyboard", "cell phone", "microwave", "oven", "toaster", "sink",
    "refrigerator", "book", "clock", "vase", "scissors", "teddy bear", "hair drier",
    "toothbrush", "hat", "jacket"
]
`

For this experiment, I added 2 new labels: "hat" and "jacket" and
created more than 20 samples of "hat" and 'jacket" in training, and more than 10 samples of these in validation.
So, I'm pretty sure the training and validation sample size is enough for training and validation.

And, the result is the following:

![bus](https://github.com/ultralytics/ultralytics/assets/132395996/5185c68b-33dd-46ee-ad82-34d549e327b0)

So, the trained model didn't get a transferred knowledge from the pretrained model (yolov8s.pt), rather the trained model replaces all the parameter values with the new data (hat and jacket).

Is there a concrete example with a few sample train and validation datasets for the transfer learning with a pretrained model?
I've been trying to make this work, but nothing works so far.

Thanks for any suggestions and comments.

**Comment by glenn-jocher** [comment]
@jshin10129 hey there! 👋

Thanks for reaching out and providing detailed insight into your setup. It seems like you've made a great effort in customizing your dataset, and adding "hat" and "jacket" as new classes should ideally work well with YOLOv8’s transfer learning capabilities.

Given your situation, a couple of tweaks might help:
- **Freezing Layers**: Instead of `freeze=[1-30]`, you might want to progressively freeze fewer layers to allow more of the model to adjust to your new classes. Sometimes, freezing too many layers can hinder the model's ability to learn new features effectively.
- **Learning Rate**: Adjusting the learning rate (`lr0`) can often help. Starting with a smaller learning rate might prevent overwriting the pretrained knowledge too rapidly.
- **Data Augmentation**: If not already done, incorporating diverse data augmentation can immensely help the model generalize better across the new classes.

Here's a slightly tweaked command considering the above points:

```bash
yolo detect train epochs=100 data=my_dataset.yaml model=yolov8s.pt imgsz=640 freeze=[1-15] lr0=0.001
```

Regarding adding another fully-connected layer, YOLOv8's architecture is designed to be quite flexible and efficient straight out of the box for a variety of tasks without the need for additional layers for new classes. 

As for a concrete example, unfortunately, we don't have a direct example with a few sample datasets specifically for the scenario you've described. But, the community keeps sharing their experiences and setups, so it might be worth keeping an eye on [Ultralytics Discussions](https://github.com/ultralytics/ultralytics/discussions) for any similar use cases that pop up! 🚀

Stay curious, and don't hesitate to experiment with the suggestions. Keep us posted on your progress; we’re here to help!

**Comment by chaitanya-srikanth** [comment]
> @chaitanya-srikanth hey there! 👋
> 
> It sounds like you're facing a challenge with transfer learning and introducing a new class to your pretrained model. When you're freezing layers and adding a new class, it's crucial to ensure that the later, unfrozen layers of the model have enough capacity to learn the new class' features.
> 
> Here's a quick tip: Instead of freezing a fixed number of initial layers, you might want to experiment with freezing layers based on the model architecture. For example, in YOLO models, you could freeze up to the last convolutional layer before the YOLO layers start. This way, the model retains its ability to detect previously learned classes while adapting to the new class with the unfrozen layers.
> 
> ```shell
> # Example of specifying which layers to freeze in a hypothetical training command
> yolo detect train data=my_dataset.yaml model=my_pretrained_model.pt freeze=[0-15]
> ```
> 
> Also, ensure your new dataset with the added class is well-represented and balanced. Sometimes, adding more varied examples of the new class can significantly help the model learn better.
> 
> Keep experimenting with the number of layers you freeze and the composition of your dataset. Sometimes, finding the right balance requires a bit of trial and error. 🚀
> 
> Let us know how it goes!

Hey @glenn-jocher. Thanks for the support, I went through the training logs and found out the issue was with the bounding box annotations. It has been resolved and the transfer learning on YoloV8 works. 
Thank you :)

**Comment by glenn-jocher** [comment]
Hey @chaitanya-srikanth, that's fantastic news! 🎉 Glad to hear the issue was sorted out and you're on track with the transfer learning on YOLOv8. Great job troubleshooting the bounding box annotations! If you ever run into another hiccup or have more insights to share, feel free to drop us a line. Happy detecting! 😊

**Comment by jshin10129** [comment]
Hi @glenn-jocher,

Thanks for the specific guidance about how to do transfer learning on Yolov8 model.
Well, so far no luck on this training still.

I've tried basically combinations of the parameters you mentioned above: Freezing Layers, Learning Rate, and Epoch numbers.
For example, (freeze=[1-15] lr0=0.001, and epochs=100), (freeze=[1-10] lr0=0.001, and epochs=100), (freeze=[1-10] lr0=0.0005, and epochs=100), (freeze=[1-10] lr0=0.001, and epochs=30), (freeze=[1-15] lr0=0.0001, and epochs=20).

The training works perfect on the new labels only (hat, jacket) when I do (freeze=[1-15] lr0=0.003, and epochs=100), and I get very good results to detect hat and jacket.
However, when I reduce the epoch number and learning rate -- expecting to balance the new labels and old labels -- it did not learn new labels and the detection accuracy became really low so it produced some gibberish outputs, such as detected 'bus' as 'oven', etc.  Among dozens of combinations for training I couldn't get any meaningful outputs -- e.g, new labels(jacket,hat) are detected along with the old labels like (bus, human, etc. up to 80 labels as in the original labels used for training the pretrained model, yolov8s.pt).

Therefore, at this point, I'm thinking fundamentally if this 'transfer learning' on Yolov8 is possible or not.
Or, are there any more parameters (except learning rate, frozen layers, and epochs) to adjust to get this achieved?

Thanks for any comments in advance.

**Comment by jshin1012** [comment]
I still haven't found out what's the combination / method to get the original labels (80 classes trained in the pretrained model), along with the new labels (e.g., hat and jacket) using transfer learning on YoloV8 -- freezing layers? or is there a method to add one more layer? or do I have to re-train some of the examples used for the pre-trained model?  
If anybody has any ideas or some experiences, please comment on this issue.
Thanks!

**Comment by glenn-jocher** [comment]
@jshin1012 hey there! 👋

Achieving a good balance in transfer learning, especially when adding new classes while maintaining the performance on the original ones, can indeed be challenging. It's a mix of getting the right training data, model adjustments, and hyperparameters just right.

For your case with YOLOv8:
- **Freezing layers** is a good strategy. You might want to play around with exactly how many to freeze. Sometimes less is more.
- **Adding an entirely new layer** specifically for new classes is generally not necessary with YOLO architectures since the models are designed to be quite adaptable to new classes as-is.
- **Re-training on some original data** along with the new class examples can help reinforce the original classes during the learning process. This way, the model doesn't "forget" the old classes while learning the new ones.

A typical training command might look something like this:

```bash
yolo detect train data=my_dataset.yaml model=yolov8s.pt epochs=100 imgsz=640 freeze=[1-15] lr0=0.001
```

Adjusting these settings to find what works best for your specific dataset might take some experimentation. Keep at it, and feel free to share your findings or further questions with the community! 🚀
